Day-8_Objects/objects.js

```diff
@@ -0,0 +1,417 @@
+    addIncome(description, amount) {
+        this.incomes.push({ description, amount });
+    },
+
+    addExpense(description, amount) {
+        this.expenses.push({ description, amount });
+    },
+
+    accountBalance() {
+        return this.totalIncome() - this.totalExpense();
+    }
+};
+
+
+const usersLevel3 = [
+    {
+        _id: 'ab12ex',
+        username: 'Alex',
+        email: 'alex@alex.com',
+        password: '123123',
+        createdAt:'08/01/2020 9:00 AM',
+        isLoggedIn: false
+    },
+    {
+        _id: 'fg12cy',
+        username: 'Asab',
+        email: 'asab@asab.com',
+        password: '123456',
+        createdAt:'08/01/2020 9:30 AM',
+        isLoggedIn: true
+    },
+    {
+        _id: 'zwf8md',
+        username: 'Brook',
+        email: 'brook@brook.com',
+        password: '123111',
+        createdAt:'08/01/2020 9:45 AM',
+        isLoggedIn: true
+    },
+    {
+        _id: 'eefamr',
+        username: 'Martha',
+        email: 'martha@martha.com',
+        password: '123222',
+        createdAt:'08/01/2020 9:50 AM',
+        isLoggedIn: false
+    },
+    {
+        _id: 'ghderc',
+        username: 'Thomas',
+        email: 'thomas@thomas.com',
+        password: '123333',
+        createdAt:'08/01/2020 10:00 AM',
+        isLoggedIn: false
+    }
+    ];
+
+// Imagine you are getting the above users collection from a MongoDB database. 
+// a. Create a function called signUp which allows user to add to the collection. If user exists, inform the user that he has already an account.
+const dmitry = {    
+    username: 'Dmitry',
+    email: 'dmitry@dmitry.com',
+    password: '121212'
+}
+const alex = {
+    _id: 'ab12ex',
+    username: 'Alex',
+    email: 'alex@alex.com',
+    password: '123123',
+    createdAt:'08/01/2020 9:00 AM',
+    isLoggedIn: false
+}
+
+function generateUniqueId(users) {  // Help function to generate uniqe ID for new user
+    const characters = 'abcdefghijklmnopqrstuvwxyz0123456789'
+    let id = ''
+
+    do {
+        id = ''
+        for (let i = 0; i < 6; i++) {
+            id += characters.charAt(Math.floor(Math.random() * characters.length))
+        }        
+    } while (users.some(user => user._id === id))
+
+    return id
+}
+
+
+function formatDate(date) {    // Help function to format date
+    let hour = date.getHours()
+    let ampm = hour >= 12 ? 'PM' : 'AM'
+    
+    hour = hour % 12
+    hour = hour? hour : 12    
+    
+    let formatYYYY = date.getFullYear()
+    let formatMM = (date.getMonth() + 1).toString().padStart(2, '0')
+    let formatDD = date.getDate().toString().padStart(2, '0')
+    let formatHH = hour.toString().padStart(2, '0')
+    let formatmm = date.getMinutes().toString().padStart(2, '0')    
+    
+    return `${formatMM}/${formatDD}/${formatYYYY} ${formatHH}:${formatmm} ${ampm}` 
+}
+
+function signUp(users, newUser) {    
+    const userExists = users.some(user => user.username === newUser.username || user.email === newUser.email)
```

Comment:
It's better to move `users.some()` to another function so that `signUp` function follows Single Responsibility principle.